### ISSUE TITLE
Expand "Say all with review" to read beyond the current object

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#19539]
### Is your feature request related to a problem? Please describe.

Say all with revie (NumpadPlus / NVDA+shift+a) is described as "Reads from the current position of the review cursor, moving it as it goes".

This function ONLY reads within the current navigator object.  While there is logic to this, practically, it could be useful to have the ability to read from this point onwards - beyond the current object.

One potentially useful use case is having NVDA set so that the review cursor follows the mouse, then a partially sighted user can move the mouse over text they want to read, and press numpad plus to have NVDA start reading at that point.  This does not work on the web, as it will only read up to the next element (the next object).

### Describe the solution you'd like

Expand numpad plus / NVDA+shift+a to start reading at the current review cursor as now, but to keep reading to the end of the document (following the path the "move to next object in flattened review" NVDA+numpad 3 / NVDA+shift+] would take through the document.

### Describe alternatives you've considered

My first thought was to remap the existing command to shift+numpad plus, but the laptop command already uses shift.

Perhaps remap the existing command to control+numpad plus / NVDA+control+shift+a for those who still want the old functionality?

Another alternative could be to leave the current command but add control to expand it to say all (beyond the current object).  That would match the logic of, for instance, arrow keys moving by letter, but control+arrow moving by word.

### Additional context

The ability to hover the mouse over text and press a key to say all is something we are frequently asked about for low vision users, and with the development of magnifier functionality, this will only become a more common request.

If there is a lot of demand for the current functionality specifically (NOT wanting the command to read beyond the current object) then perhaps using control+numpad plus / NVDA+control+shift+a to add the new functionality would be preferred.  If, however, users feel that the proposed functionality is an improvement, then I would expand the current keystroke (and adding a new one, assigned or unassigned) for the old functionality.  I considered whether to make this a discussion, but I feel we will likely get a feeling fairly quickly from reactions to this issue.

**Comment by CyrilleB79** [comment]
It seems that this issue first suggests a solution (in "Is your feature request related to a problem? Please describe.") and then explains the real problem a user is facing (in "Additional context").

I'd recommend to re-write the initial description with the real root problem described in section "Is your feature request related to a problem?" and then come with a possible solution in the next section.

To me, the real original problem is that a visually impaired user can see some text on the screen (or a piece of it) and wants the whole text to be read. The easiest way would be to point with the mouse and to execute a command to read all from there (ideally, `NVDA+leftMouseClick`). This has been mentioned in https://github.com/nvaccess/nvda/issues/19495#issuecomment-3788993859 and maybe elsewhere).

In your current initial description, you suggest to use the review cursor for this and not the system or browse mode cursor. Why? Have you also investigated what Narrator or Windows Magnifier does? I guess they have a similar feature.

**Comment by Qchristensen** [comment]
The mouse user scenario is one of the main ones I can immediately think of where this would be useful, it is definitely not the only one - I have got somewhere with the review cursor with the keyboard before and wanted to read on past the current object.  Theoretically you can move the focus to the location and use regular say all (if you can move focus there of course), but that requires a couple of extra steps, and doesn't always work (particularly on the web).

The other reason for the mouse scenario is that you can already have the review cursor follow the mouse, so this would be another way of accomplishing the "move mouse to text, press a keystroke to read" functionality that low vision users ask for.

Perhaps a question to ask is - however you get there, when you use numpad plus or NVDA+shift+a to say all with review - what are you reading and what do you WANT to read?

- If you've navigated to an object and specifically want to know exactly (and only) the text in this object, then the current functionality is likely what you want.
- If you have got to an object and want to read "from this point", then the expanded functionality suggested might be more useful.

And if you DON'T use this function at all is it because you find it limiting as it currently is, or because you prefer to move word or line by line at this point, or something else entirely?

**Comment by CyrilleB79** [comment]
> Theoretically you can move the focus to the location and use regular say all (if you can move focus there of course), but that requires a couple of extra steps, and doesn't always work (particularly on the web).

Doesn't `NVDA+shift+pavNumMinus` double press work on the web? It seems to work on my end. But if it doesn't in some specific cases, it should be reported as an issue (if not already).

**Comment by Qchristensen** [comment]
It does seem to work - I retract that :)  What threw me was needing to press NVDA+numpad multiply to navigate to the object under the mouse first, which took me awhile to work out because I was thinking I already had set the review cursor to follow the mouse - so there might be a separate issue there (or I just need to get my head around it better) - but either way to "move the mouse to something and read it with (regular) say all" currently requires:

1) Hover the mouse over the desired text
2) Press NVDA+numpad multiply to navigate to the object under the mouse
3) Press NVDA=shift+numpad minus twice to move focus to the review cursor
4) Press NVDA+down arrow to read.

Which isn't exactly a simple solution.  The proposal in this issue would instead make that:

1) Hover the mouse over the desired text
2) Press numpad plus

Alternatively being able to move the mouse and click to read (#19495) or have the system focus / caret follow the mouse instead of the review cursor (so you can hover over the text then press NVDA+down arrow / NVDA+l  - as suggested in #19540) are alternatives to the mouse user scenario - all three with their pros and cons.

I do think there is merit to this idea even beyond mouse users, but happy to be overruled if no-one else thinks that :)

**Comment by seanbudd** [comment]
This issue needs more information on the problem it is trying to solve including user stories

**Comment by Qchristensen** [comment]
I would be interested in hearing from keyboard-only users who use say all with review - and whether they would find this better overall (which would lean to changing the current behaviour), or only in some situations (which would lean towards ensuring there is a keystroke for each functionality).

Since #19540 was not feasible, I think the mouse user example is a sufficient use case for this issue regardless:  Low vision mouse users need a way of moving the mouse over text, and pressing a single keystroke to have it read from that point onward.

#4896 is currently being considered with the preferred implementation of holding the NVDA key and clicking.  That is good for some users as clicking is relatively intuitive, and combining with the NVDA key will prevent accidental activation of controls you are trying to read.  HOWEVER, that solution will be problematic for some users.  For some low vision users, accurately clicking with the mouse can be problematic.  So letting go of the mouse once they hear what they want to read and pressing a key on the keyboard will be more reliable.  If we add in the need to use two hands - to press the NVDA key AND a mouse button, that will make it more difficult for some of those users - And I can't find a technical page on it, but a Google search indicates sticky keys will NOT work when you need to hold a key and click the mouse.  Expanding say all with review would resolve this for low vision mouse users who also require sticky keys.
